art/estimators/certification/interval/pytorch.py

```diff
@@ -0,0 +1,376 @@
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements certification using interval (box) domain certification.
+
+| Paper link: https://ieeexplore.ieee.org/document/8418593
+"""
+
+from typing import List, Optional, Tuple, Union, Any, TYPE_CHECKING
+
+import logging
+import warnings
+
+import numpy as np
+import torch
+
+from art.estimators.certification.interval.interval import (
+    IntervalConv2D,
+    IntervalDenseLayer,
+    IntervalReLU,
+    IntervalBounds,
+)
+from art.estimators.classification.pytorch import PyTorchClassifier
+
+
+if TYPE_CHECKING:
+    from art.utils import CLIP_VALUES_TYPE, PREPROCESSING_TYPE
+    from art.defences.preprocessor import Preprocessor
+    from art.defences.postprocessor import Postprocessor
+
+
+class ConvertedModel(torch.nn.Module):
+    """
+    Class which converts the supplied pytorch model into an equivalent model
+    which uses abstract operations
+    """
+
+    def __init__(self, model: "torch.nn.Module", channels_first: bool, input_shape: Tuple[int, ...]):
+        super().__init__()
+        modules = []
+        self.interim_shapes: List[Tuple] = []
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        self.forward_mode: str
+        self.forward_mode = "abstract"
+        self.reshape_op_num = -1
+
+        # pylint: disable=W0613
+        def forward_hook(input_module, hook_input, hook_output):
+            modules.append(input_module)
+            self.interim_shapes.append(tuple(hook_input[0].shape))
+
+        for module in model.children():
+            module.register_forward_hook(forward_hook)
+
+        if channels_first:
+            input_for_hook = torch.rand(input_shape).to(self.device)
+        else:
+            raise ValueError("Please provide data in channels first format")
+
+        input_for_hook = torch.unsqueeze(input_for_hook, dim=0)
+        model(input_for_hook)  # hooks are fired sequentially from model input to the output
+
+        self.ops = torch.nn.ModuleList()
+        for module, shapes in zip(modules, self.interim_shapes):
+            weights_to_supply = None
+            bias_to_supply = None
+            print("registered", type(module))
```

Comment:
Can we either remove this line or convert to logging?

Reply:
Can be removed - it was primarily for sanity checking during development